sdks/java/core/src/main/java/org/apache/beam/sdk/io/Read.java

```diff
@@ -375,4 +422,361 @@ public void checkDone() throws IllegalStateException {
+      // been created.
+      if (restriction.getValue() != null
+          && !(restriction.getValue()
+              instanceof UnboundedSource.CheckpointMark.NoopCheckpointMark)) {
+        receiver.output(restriction);
+      }
+
+      try {
+        for (UnboundedSource<OutputT, CheckpointT> split :
+            restriction.getKey().split(DEFAULT_DESIRED_NUM_SPLITS, pipelineOptions)) {
+          receiver.output(KV.of(split, null));
+        }
+      } catch (Exception e) {
+        receiver.output(restriction);
+      }
+    }
+
+    @NewTracker
+    public RestrictionTracker<
+            KV<UnboundedSource<OutputT, CheckpointT>, CheckpointT>, UnboundedSourceValue<OutputT>[]>
+        restrictionTracker(
+            @Restriction KV<UnboundedSource<OutputT, CheckpointT>, CheckpointT> restriction,
+            PipelineOptions pipelineOptions) {
+      return new UnboundedSourceAsSDFRestrictionTracker(restriction, pipelineOptions);
+    }
+
+    @ProcessElement
+    public ProcessContinuation processElement(
+        RestrictionTracker<
+                KV<UnboundedSource<OutputT, CheckpointT>, CheckpointT>, UnboundedSourceValue[]>
+            tracker,
+        OutputReceiver<ValueWithRecordId<OutputT>> receiver,
+        BundleFinalizer bundleFinalizer)
+        throws IOException {
+      UnboundedSourceValue<OutputT>[] out = new UnboundedSourceValue[1];
+      while (tracker.tryClaim(out)) {
+        receiver.outputWithTimestamp(
+            new ValueWithRecordId<>(out[0].getValue(), out[0].getId()), out[0].getTimestamp());
+      }
+
+      // Add the checkpoint mark to be finalized if the checkpoint mark isn't trivial.
+      KV<UnboundedSource<OutputT, CheckpointT>, CheckpointT> currentRestriction =
+          tracker.currentRestriction();
+      if (currentRestriction.getValue() != null
+          && !(tracker.currentRestriction().getValue() instanceof NoopCheckpointMark)) {
+        bundleFinalizer.afterBundleCommit(
+            Instant.now().plus(Duration.standardMinutes(DEFAULT_BUNDLE_FINALIZATION_LIMIT_MINS)),
+            currentRestriction.getValue()::finalizeCheckpoint);
+      }
+
+      // If we have been split/checkpoint by a runner, the tracker will have been updated to the
+      // empty source and we will return stop. Otherwise the unbounded source has only temporarily
+      // run out of work.
+      if (tracker.currentRestriction().getKey() instanceof EmptyUnboundedSource) {
+        return ProcessContinuation.stop();
+      }
+      return ProcessContinuation.resume();
+    }
+
+    @GetRestrictionCoder
+    public Coder<KV<UnboundedSource<OutputT, CheckpointT>, CheckpointT>> restrictionCoder() {
+      return KvCoder.of(
+          SerializableCoder.of(new TypeDescriptor<UnboundedSource<OutputT, CheckpointT>>() {}),
+          NullableCoder.of(restrictionCoder));
+    }
+
+    /**
+     * A named tuple representing all the values we need to pass between the {@link UnboundedReader}
```

Comment:
Is "named tuple" also java terminology, or should this be a POJO?

Reply:
not really, borrowed it from python, will fix.